sda-commons-server-kafka/src/main/java/org/sdase/commons/server/kafka/producer/MetadataContextAwareKafkaProducer.java

```diff
@@ -0,0 +1,145 @@
+ *
+ * @param <K> the type of the message key
+ * @param <V> the type of the message value
+ */
+public class MetadataContextAwareKafkaProducer<K, V> implements Producer<K, V> {
+
+  private final Producer<K, V> delegate;
+
+  private final Set<String> metadataFields;
+
+  /**
+   * @param delegate the actual {@link Producer} that is used to interact with the Kafka cluster.
+   * @param metadataFields the configured fields that are used as metadata
+   */
+  public MetadataContextAwareKafkaProducer(Producer<K, V> delegate, Set<String> metadataFields) {
+    this.delegate = delegate;
+    this.metadataFields = Optional.ofNullable(metadataFields).orElse(Set.of());
+  }
+
+  @Override
+  public void initTransactions() {
+    delegate.initTransactions();
+  }
+
+  @Override
+  public void beginTransaction() throws ProducerFencedException {
+    delegate.beginTransaction();
+  }
+
+  @Override
+  @SuppressWarnings("deprecation")
+  public void sendOffsetsToTransaction(
+      Map<TopicPartition, OffsetAndMetadata> offsets, String consumerGroupId)
+      throws ProducerFencedException {
+    delegate.sendOffsetsToTransaction(offsets, consumerGroupId);
+  }
+
+  @Override
+  public void sendOffsetsToTransaction(
+      Map<TopicPartition, OffsetAndMetadata> offsets, ConsumerGroupMetadata groupMetadata)
+      throws ProducerFencedException {
+    delegate.sendOffsetsToTransaction(offsets, groupMetadata);
+  }
+
+  @Override
+  public void commitTransaction() throws ProducerFencedException {
+    delegate.commitTransaction();
+  }
+
+  @Override
+  public void abortTransaction() throws ProducerFencedException {
+    delegate.abortTransaction();
+  }
+
+  @Override
+  public Future<RecordMetadata> send(ProducerRecord<K, V> producerRecord) {
+    return delegate.send(withMetadataContext(producerRecord));
+  }
+
+  @Override
+  public Future<RecordMetadata> send(ProducerRecord<K, V> producerRecord, Callback callback) {
+    return delegate.send(withMetadataContext(producerRecord), callback);
+  }
+
+  @Override
+  public void flush() {
+    delegate.flush();
+  }
+
+  @Override
+  public List<PartitionInfo> partitionsFor(String topic) {
+    return delegate.partitionsFor(topic);
+  }
+
+  @Override
+  public Map<MetricName, ? extends Metric> metrics() {
+    return delegate.metrics();
+  }
+
+  @Override
+  public void close() {
+    delegate.close();
+  }
+
+  @Override
+  public void close(Duration timeout) {
+    delegate.close(timeout);
+  }
+
+  private ProducerRecord<K, V> withMetadataContext(ProducerRecord<K, V> producerRecord) {
+    if (metadataFields.isEmpty()) {
+      return producerRecord;
+    }
+    var headers = new RecordHeaders(producerRecord.headers());
+    MetadataContext metadataContext = MetadataContext.current();
+    for (String metadataField : metadataFields) {
+      List<String> valuesByKey = metadataContext.valuesByKey(metadataField);
+      if (valuesByKey == null) {
```

Comment:
What about `valuesByKey.isEmpty()`?
Do we want to support that case at all? We could also change the API for the metadata context and only store keys that have at least one value.

Reply:
> Do we want to support that case at all? We could also change the API for the metadata context and only store keys that have at least one value.

Better safe than sorry. In the end it is an interface that may be implemented different.

> What about valuesByKey.isEmpty()?

That's just to avoid NPEs. Is it so much more expensive to iterate over an empty stream than checking for empty? In the current implementation, the code has less branches is is imho easier to read.